stable_baselines/common/distributions.py

```diff
@@ -344,12 +352,17 @@ def __init__(self, nvec, flat):
         """
         self.flat = flat
         self.categoricals = list(map(CategoricalProbabilityDistribution, tf.split(flat, nvec, axis=-1)))
+        self.action_mask_vector = action_mask_vector
 
     def flatparam(self):
         return self.flat
 
     def mode(self):
-        return tf.cast(tf.stack([p.mode() for p in self.categoricals], axis=-1), tf.int32)
+        modes = []
+        for idx, p in enumerate(self.categoricals):
```

Comment:
So, I do element wise multiplication instead of boolean_mask because I don't exactly understand how tf.boolean_mask applies to a batch of actions.

The way I originally had it, was in tf.boolean_mask(self.logits, self.action_mask_vector[0]), which worked for my env but I didn't know how it behaved with other envs. 

This approach will work if you relu the output instead of TanH. If you tanh, and all your valid actions have negative values, the element wise multiplication will result in an invalid action being selected because 0 > {all negative valued valid actions}.

@araffin

Reply:
I have to think a bit more for that one...